charts/kyverno/templates/NOTES.txt

```diff
@@ -1,19 +1,23 @@
-Thank you for installing {{ .Chart.Name }} {{ .Chart.Version }} 😀
+Chart version: {{ .Chart.Version }}
+Kyverno version: {{ default .Chart.AppVersion (default .Values.image.tag .Values.initImage.tag) }}
 
-Your release is named {{ .Release.Name }}, app version {{ default .Chart.AppVersion (default .Values.image.tag .Values.initImage.tag) }}
+Thank you for installing {{ .Chart.Name }}! Your release is named {{ .Release.Name }}.
 
 {{- if not .Values.replicaCount }}
-WARNING: Setting replicas count below 3 means Kyverno is not running in high availability mode.
+⚠️  WARNING: Setting replicas count below 3 means Kyverno is not running in high availability mode.
 {{- else if lt .Values.replicaCount 3 }}
-WARNING: Setting replicas count below 3 means Kyverno is not running in high availability mode.
+⚠️  WARNING: Setting replicas count below 3 means Kyverno is not running in high availability mode.
 {{- end }}
 
-{{- if empty .Values.config.webhooks }}
-WARNING: Kyverno has been installed with no Namespace exclusions which may prevent normal cluster operations in cluster or Node failure scenarios. Please see the documentation at https://kyverno.io/installation/#security-vs-operability to understand the risks.
-{{- else }}
-WARNING: Kyverno has been installed and some Namespaces and/or resources have been excluded from webhooks. Please see the documentation at https://kyverno.io/installation/#security-vs-operability to understand the risks.
+{{- if .Values.config.webhooks }}
+  {{- if has (include "kyverno.namespace" .) (first (first .Values.config.webhooks).namespaceSelector.matchExpressions).values }}
```

Comment:
How about `matchLabels` ?

Reply:
A label selector comes with two fields: `matchLabels` and `matchExpressions`, both can be set and are ANDed together (see https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements).

In the logic above, the check is not reliable.

IMHO we should not try to understand what the label selectors mean, it's too complicated.

Reply:
We only define in the values file `matchExpressions` and I have added the label key as `kubernetes.io/metadata.name` which is an immutable label the API server automatically assigns to each Namespace with the value equal to that Namespace name. Therefore, it is a pretty reliable way to determine if the Namespace into which Kyverno is deployed is also being excluded.

Reply:
https://kubernetes.io/docs/concepts/overview/working-with-objects/_print/#automatic-labelling

Reply:
The user can set whatever he wants, what if I use the values below ?
```yaml
config:
  webhooks:
  - namespaceSelector:
      matchLabels:
        kubernetes.io/metadata.name: kyverno
```
Or something like:
```yaml
config:
  webhooks:
  - namespaceSelector:
      matchExpressions:
        - key: kubernetes.io/metadata.name
          operator: In
          values:
            - something
```
Both examples exclude the `kyverno` ns.
Your logic only works with your example ;)

Reply:
Also, we can have multiple expressions:
```yaml
config:
  webhooks:
  - namespaceSelector:
      matchExpressions:
        - key: kubernetes.io/metadata.name
          operator: NotIn
          values:
            - kube-system
            - kyverno
        - key: foo
          operator: NotIn
          values:
            - bar
            - baz
```

Reply:
Understood. So we need a second check for `matchLabels`. If we can merge this now, I can try and make it more flexible so it accounts for both, but since we provide a default of `matchExpressions` we can at least confirm it works with what we provide. Sound good?

Reply:
- if there is an  `objectSelector` stanza you can't determine reliably (both `namespaceSelector` and `objectSelector` can be part of the same object)
- if there is only `matchLabels`, `kubernetes.io/metadata.name: kyverno` should be the only key-value present
- if there is only `matchExpressions`, all entries should use `kubernetes.io/metadata.name` as key and given the operator you should study the values
- if there is both `matchLabels` and `matchExpressions`, the two bullet points above should be true

But even with that, I could have created the kyverno ns beforehand and set a label on it, then use this label in the webhooks config, effectively making it excluded and safe. That's not something you can determine.

Reply:
Fortunately nobody reads the `NOTES` though ;-)

More seriously, I would keep the message generic enough.
Excluding resources deserves a warning, not excluding resources also deserves a warning... in all cases there is a warning to show, only the message should be different and linking to a single page that covers both should be fine.

Reply:
Bottom line it's up to you, if you are confident that we can implement a logic robust enough go ahead.
We can try and if it proves to be too hard revert to something simpler.
The most important is that the users see it and have a web page to browse to get more details.

Reply:
Ok, so are you thinking that the best we can do is what's already [here](https://github.com/kyverno/kyverno/blob/a6924a11abc82a1d47c40480a51f1b4e277889f9/charts/kyverno/templates/NOTES.txt#L11-L15) but just updating the messages?

Reply:
I think we shouldn't say something like `You didn't exclude the kyverno namespace ...` because:
- there's not a single way to exclude admission requests from a webhook (doing it with `objectSelector` is a valid approach too)
- determining if the user choice is safe or not looks impossible except for trivial cases (simple config using built in keys)

Maybe we can show the warning without conditions, I'm not sure it's worth the effort to try to be smart here.
It depends on the goal, but as being precise seems very difficult I would say that the best we can do is try to draw attention from users and warn them as much as we can, inviting them to visiting a web page explaining things in more details.

Again, it's just my personal opinion, I would really love to get something as clean and precise as possible, I just don't see how to do it.

Reply:
Ok, thanks for the thoughts. I think I'm with you. I wanted to try and be "smart" and give users (who see the notes) the best info we can, and while that may work if they uncomment our values.yaml file and replace existing values, it won't work for much beyond that making this system brittle. I think maybe simplest is best in this case. Let me alter this and ask for your review again.

Reply:
@eddycharly would appreciate your re-review.

Reply:
:+1: